### PR TITLE
Allow for overwriting metrics in set_metrics method

### PR DESCRIPTION
**Pull request by jdldeauna**
- [X] Include overwrite kwarg in set_metrics
 - [X] Tests added in test_metrics/test_set_metric_overwrite_true and test_set_metric_overwrite_false

### CODE REVIEW

File: xgcm/grid.py
Comment (jdldeauna) [inline xgcm/grid.py:1427]: I think this functionality will be implemented once interp_like is combined with get_metric.

File: xgcm/grid.py
Comment (jbusecke) [inline xgcm/grid.py:1289]: This should only be triggered if overwrite is true? I think you just need to move it a line down under `if overwrite:`. Could you add this to the tests,  and check if this is triggered only if the value is actually overwritten?

File: xgcm/grid.py
Comment (jbusecke) [reply xgcm/grid.py:1427]: yes lets keep this in another PR.

File: xgcm/test/test_metrics.py
Comment (jbusecke) [inline xgcm/test/test_metrics.py:319]: I might split this into two tests: One for overwrite=True and another for overwrite=False. Each of these could test that the appropriate warning/error is raised each time (see my comment above). We can discuss this later too.

File: xgcm/test/test_metrics.py
Comment (jbusecke) [inline xgcm/test/test_metrics.py:337]: Is this actually overwriting anything? I dont think you are setting the metric before?

To really test that a new dataarray is set I would do the following in the test:
1) Set all metrics that are in the dataset with `grid=Grid(ds, ..., metrics=metrics)`
2) Modify a metric and assign it under a new name. For example:
```python
metric_varname_modified = f'{metric_varname}_overwrite'
ds = ds.assign_coords(metric_varname_modified, ds[metric_varname]*10)
```
3) Then assign this modified metric to the dataset with `grid.set_metric(metric_axes, metric_varname_modified)`. You will have to write this in a loop, this is just an example. 
4) Now you can check explicitly that the new metric is actually 10 times as large. Without that you cant be really sure that the metric has been replaced!

**Review by jbusecke — CHANGES_REQUESTED** [review]
Thanks for this @jdldeauna. As far as I can see the code changes look good, but I think the tests need some modifications to ensure the behavior we want. Happy to help further with this if anything is unclear.

File: xgcm/grid.py
Comment (jdldeauna) [reply xgcm/grid.py:1289]: So right now the flow is: 
1) check if there is input which matches an existing metric (Line 1284)
2) cycle through existing metrics to find the match 
3) check if the dimensions between input and existing metric match 
4) replace if the value of overwrite = True
Just to clarify would you want to switch steps 3 and 4?

Comment (jbusecke) [reply xgcm/grid.py:1289]: Oh no. I am just concerned about the warning. 

Currently you have somthing like this:
```python
if set(value_new.dims) == set(ve.dims):
    warn(... replacing...).# this will be triggered regardless of whether overwrite is true or false
    if overwrite:
        # replace metric
    else:
       raise error
```

What I suggest is simply to move the warning into the if statement. Otherwise you will get a very confusing output for overwrite false. It will basically say: "I am overwriting" (which it isnt) and then fail.

Comment (jbusecke) [reply xgcm/grid.py:1289]: What I would do here is again start with the test: Make a test that checks that this warning is raised when overwrite=True, and more importanty: Make sure it is **not** raised when overwrite=False. I believe this will fail with the current code, and then pass when you move the warning.

File: xgcm/test/test_metrics.py
Comment (jdldeauna) [reply xgcm/test/test_metrics.py:337]: Hi, I was trying to incorporate this into the tests, and I wanted to clarify. Is the purpose of the test to:
1. Add a new metric that is 10x the value of the old one (e.g., you have both `dx_t` and `dx_t_overwrite` as metrics)
2. Replace an existing metric with something named similarly but 10x in value (e.g., you end up with one `dx_t` metric that's 10x in value as the original one from metrics=metrics)

Comment (jbusecke) [reply xgcm/test/test_metrics.py:337]: Yes that way you can be sure that what you are checking is not just the metric set in the first place.

File: xgcm/test/test_metrics.py
Comment (jdldeauna) [inline xgcm/test/test_metrics.py:367]: I couldn't integrate this test with `test_set_metric_replace` because if metrics are set as `metrics=metrics` from the beginning, I can't append a new metric when its dimensions already matches an existing metric (can only overwrite).

File: xgcm/test/test_metrics.py
Comment (jdldeauna) [inline xgcm/test/test_metrics.py:337]: Hacky approach to using `metric_axes = ("X","Y")`😅  
Feel free to suggest an alternative!

File: xgcm/test/test_metrics.py
Comment (jbusecke) [reply xgcm/test/test_metrics.py:367]: You dont have to set all the metrics in the beginning. How about you modify the `metrics` input dict according to your input parameters:
```python
metric_names = ['dx_t', 'dx_ne'] #not sure this is the right variable name, just a sketch
ds, coords, metrics = ....
metrics = {k:[m for m in v if m in metric_names] for k,v in metrics.items()} # this will then only have the above names in it.
grid = Grid(..., metrics=metrics, ...)
```

File: xgcm/test/test_metrics.py
Comment (jbusecke) [reply xgcm/test/test_metrics.py:337]: does `frozenset(list(metric_axes))` work in both cases?

Comment (jdldeauna) [reply xgcm/test/test_metrics.py:337]: It worked!

File: xgcm/grid.py
Comment (jbusecke) [inline xgcm/grid.py:1304]: These few lines of code are repeated above. Can we factor them out into a small helper method? Something like:
```
def _maybe_get_var(self, varname):
     if varname not in self._ds:
         raise KeyError(f"Variable {varname} not found in dataset.")
    else:
        return self._ds[varname].reset_coords(drop=True)
```
and then call it as `metric_var = self._maybe_get_var(metric_varname)` in both cases?

File: xgcm/test/test_metrics.py
Comment (jbusecke) [inline xgcm/test/test_metrics.py:342]: I ❤️ this test! Great job.

**Review by jbusecke — APPROVED** [review]
Hey @jdldeauna, this is in great shape! I had one (optional) suggestion to make the code even cleaner, but I would also be happy to merge this as is!

File: xgcm/grid.py
Comment (dcherian) [inline xgcm/grid.py:1281]: could remove this else and avoid one level of indentation. It will only run if the `if` condition is not True.

File: xgcm/grid.py
Comment (dcherian) [inline xgcm/grid.py:1296]: `metric_varname` is the "new" metric the user is trying to assign. I think you want to print `ve.name` instead.


```suggestion
                                    f"Metric variable {ve.name} with dimensions {ve.dims} already assigned in metrics."
                                    f" Overwrite {ve.name} with {metric_varname} by setting overwrite=True."
                                )
```

File: xgcm/grid.py
Comment (dcherian) [inline xgcm/grid.py:1272]:
```suggestion
        existing_metric_axes = set(self._metrics.keys())
```

should also work? `dict.keys()` and `dict.values()` are "iterables" so you can cast them to `tuple` `set` and `list` directly.

File: xgcm/grid.py
Comment (dcherian) [inline xgcm/grid.py:1301]:
```suggestion
            # no metrics for metric_axes yet; initialize empty list
```

I was confused about when this else is being triggered. Another reason to reduce indentation above,

File: xgcm/grid.py
Comment (dcherian) [reply xgcm/grid.py:1304]: Alternatively we can just make sure that  all `metric_varname` entries are `in self._ds.variables` right at the top before we loop through so we only ever write this code once.

`.variables` includes both `.data_vars` and `.coords`

File: xgcm/test/test_metrics.py
Comment (dcherian) [reply xgcm/test/test_metrics.py:342]: agreed! :heart:

File: xgcm/test/test_metrics.py
Comment (dcherian) [inline xgcm/test/test_metrics.py:348]: haha list comprehension in a dict comprehension. dang!

File: xgcm/test/test_metrics.py
Comment (dcherian) [inline xgcm/test/test_metrics.py:369]:
```suggestion
def test_set_metric_errors():
```

We should also test `grid.set_metrics("X", "foo")` raises an error.

What should `grid.set_metrics("X", "dx_e", overwrite=False)` do? It's identical to an already set metric.

**Review by dcherian — COMMENTED** [review]
looking great.

File: xgcm/test/test_metrics.py
Comment (jbusecke) [reply xgcm/test/test_metrics.py:348]: 👾

File: xgcm/test/test_metrics.py
Comment (jdldeauna) [reply xgcm/test/test_metrics.py:342]: Yehey thank you! ❤️

File: xgcm/test/test_metrics.py
Comment (jdldeauna) [reply xgcm/test/test_metrics.py:348]: 👹

File: xgcm/grid.py
Comment (dcherian) [inline xgcm/grid.py:1274]:
```suggestion
            if metric_varname not in self._ds.variables:
```

that should check all variables including those in `.coords`

**Review by dcherian — APPROVED** [review]

**Comment by dcherian** [comment]
Great nice job Diane!

go ahead and merge since you have two approvals

**Comment by jdldeauna** [comment]
> Great nice job Diane!
> 
> go ahead and merge since you have two approvals

Did a small tweak to a test before merging, thanks Deepak!

**Pull request merged**
